qa/lucene-index-compatibility/src/javaRestTest/java/org/elasticsearch/lucene/FullClusterRestartLuceneIndexCompatibilityIT.java

```diff
@@ -374,4 +387,118 @@ public void testRestoreIndexOverClosedIndex() throws Exception {
             deleteIndex(index);
         }
     }
+
+    /**
+     * 1. creates an index on N-2 and performs async_search on it that is kept in system index
+     * 2. After update to N-1 (latest) perform a system index migration step, also write block the index
+     * 3. on N, check that async search results are still retrievable and we can write to the system index
+     */
+    public void testAsyncSearchIndexMigration() throws Exception {
+        final String index = suffix("index");
+        final String asyncSearchIndex = ".async-search";
+        final int numDocs = 2431;
+
+        final Request asyncSearchRequest = new Request("POST", "/" + index + "/_async_search?size=100&keep_on_completion=true");
+
+        if (isFullyUpgradedTo(VERSION_MINUS_2)) {
+            createIndex(
+                client(),
+                index,
+                Settings.builder()
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, randomInt(2))
+                    .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true)
+                    .build()
+            );
+            indexDocs(index, numDocs);
+            ensureGreen(index);
+
+            assertThat(indexVersion(index), equalTo(VERSION_MINUS_2));
+            String asyncId = searchAsyncAndStoreId(asyncSearchRequest, "n-2_id");
+            ensureGreen(asyncSearchIndex);
+
+            checkRetrieveAsyncSearch(asyncId, numDocs);
+            assertBusy(() -> assertDocCountNoWarnings(client(), asyncSearchIndex, 1));
+            assertThat(indexVersion(asyncSearchIndex, true), equalTo(VERSION_MINUS_2));
+            return;
+        }
+
+        if (isFullyUpgradedTo(VERSION_MINUS_1)) {
+            // check .async-search index is readable
+            assertThat(indexVersion(asyncSearchIndex, true), equalTo(VERSION_MINUS_2));
```

Comment:
It is checking that the async index was created in N-2, not that it is readable. Unless the comment applies to the whole section?

Reply:
Yes it was meant for the whole section but I think its a good idea to add a check that we can read the "old" search result even before the migration later